### PR TITLE
fix the -runtime-variant option for bytecode

### PR DESCRIPTION
**Pull request by damiendoligez**
The `-runtime-variant` option has been broken since 4.03.0 because the `camlheaderd` and `camlheaderi` files are installed under the wrong names. The first commit fixes this problem.

The failure is silent and a bytecode file is produced that has the executable flag but doesn't work because the header is missing. The second commit adds some error handling for this case (in particular when the user specifies a runtime variant that doesn't exist).

Note: the first bug was introduced by myself in commit 0225ca01e39289ce1801fb09cd011cdbfb542b8d

### CODE REVIEW

**Comment by damiendoligez** [comment]
See also #2309 and #6199.

File: Changes
Comment (dra27) [inline Changes]:
```suggestion
- GPR#8537: fix the -runtime-variant option for bytecode
```
"byte-code" seems to have gone the way of "O'Caml" 🙂

File: bytecomp/bytelink.ml
Comment (dra27) [inline bytecomp/bytelink.ml]: Why the elimination of the back-tick here? FWIW, I'd prefer to see other instances of `'%s'` changed to `` `%s'`` rather than this way around?

File: bytecomp/bytelink.ml
Comment (dra27) [inline bytecomp/bytelink.ml]:
```suggestion
      fprintf ppf "System error while copying file %s: %s" header msg
```

**Review by dra27 — COMMENTED** [review]
Also related to #2267. I appear to be in a very nitty mood this morning with the other comments.

**Comment by dra27** [comment]
I have a nagging concern that the flexlink bootstrap relies on the "silent fail" behaviour when the header is missing - I just want to check what's going on there before this is merged (I'm suspicious that in bytecode-only mode, it may result in a corrupt `flexlink.exe` being installed).

File: bytecomp/bytelink.ml
Comment (dbuenzli) [reply bytecomp/bytelink.ml]: @dra27 some people from your faculty [may disagree](https://www.cl.cam.ac.uk/~mgk25/ucs/quotes.html) ;-)

**Comment by dra27** [comment]
Just to confirm - this PR will break the flexdll bootstrap build the next time someone commits a bootstrap. Do you mind if I push fixes directly?

File: bytecomp/bytelink.ml
Comment (dra27) [reply bytecomp/bytelink.ml]: Quite possibly - but let's remove them wholesale, rather than piecemeal with other changes 🙂

Comment (dra27) [reply bytecomp/bytelink.ml]: (I'd also prefer them either to be post-processed to the correct Unicode character before going to the console, or for us finally to allow the correct Unicode characters directly in the message, both of which would keep Markus happy!)

**Comment by dra27** [comment]
I've rebased and added commits to fix the flexdll bootstrap in [my fork](https://github.com/dra27/ocaml/tree/runtime-variant-fixes)

**Comment by damiendoligez** [comment]
@dra27 I've applied all your suggestions and integrated your flexdll fixes (tested by bootstrapping flexdll on mingw64).

I've also moved the Change entry to 4.09, as I intend to cherry-pick this patch.

**Review by dra27 — APPROVED** [review]

**Comment by dra27** [comment]
@damiendoligez - squash or merge, and still good for 4.09?

**Comment by gasche** [comment]
This is a bugfix so it should be good for 4.09.
(I have a marked preference for having merge commits. In this case the history is not great, but we don't want an extra roundtrip with the contributor, and it's not so bad, so I would just Merge directly.)

**Comment by dra27** [comment]
FWIW, I just squashed the Changes bit - the flexdll and other changes are a bit subtle, so seem worth retaining cleanly for bisecting

**Comment by dra27** [comment]
I'll merge and cherry-pick once CI passes

**Comment by dra27** [comment]
4.09: 617153cec62311a03d2dd4e83ff7ffd6bd0de8f1

**Comment by damiendoligez** [comment]
Thanks!

**Pull request merged**
